# Single-cell epigenomic identification of inherited risk loci in Alzheimer’s and Parkinson’s disease
https://www.biorxiv.org/content/10.1101/2020.01.06.896159v1


## Bulk ATAC-seq data processing / peak calling ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/bulk_atacseq_preprocessing)

## Differential accessibility analysis with DESeq2 and limma voom ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/differential_accessibility_analysis)

## Single-cell ATAC-seq data processing / peak calling ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/single_cell_atacseq_preprocessing)

## LD score regression: Cell type specific analyses ##
[scripts](https://github.com/lfresard/LDSC_Corces_etal_LDSC/tree/5bcf805562c72575cd2e614194c888af9a946407)

## Colocalization analysis ##
[scripts](https://github.com/mikegloudemans/alzheimers-parkinsons-colocalization/tree/29e228c1cdd68f1fd10ff84bea26b95e4faac7ba)

## SVM model training ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/svm_training)

## SNP effect scoring with SVMs ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/svm_snp_scoring)

## SNP effect interpretation ##
[notebooks](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/snp_interpretation)

## Neural network training ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/nn_training)

## Neural network interpretation/ variant score visualization ##
[scripts](https://github.com/kundajelab/alzheimers_parkinsons/tree/master/nn_interpretation)

